User-Agent choice

Be4stElectrjc opened the pull request:
Hi, using the headers passing option found an easy solution for the problem. Simply added on the fly the User-Agent during the stream parsing. From the addon settings you can choose the User-Agent you like. I didn't update all the strings yet for the new UI feature...
Keep in mind that User-Agent change it's only for the channels included in the playlist, so Kodi keeps it's own identity. I am open to discussion, let me know.

phunkyfish commented:
If you want to add this feature you first need to add it to the Matrix version of the addon. Bear in mind that the Matrix version is quite different architecturally. 

In addition the only strings.po file that should be modified en_gb. The others are generated by an automatic system.

ksooo commented:
And please use more speaking commit messages. Describe the actual change, not which file you modifies.

Be4stElectrjc commented:
Ok, will do that.

Be4stElectrjc committed:
Added variables and settings

Be4stElectrjc commented:
Reformatted all the messages etc. 
Do i need to open another PR for Matrix?

phunkyfish commented:
Yes, you will need a PR for Matrix, bear in mind the code base for the Matrix version is quite different. But we can review this PR first if you would like. Then you will have a better sense what would be required for Matrix.

phunkyfish commented on pvr.iptvsimple/resources/language/resource.language.it_it/strings.po:
```diff
@@ -78,7 +78,7 @@ msgstr "Loghi canali"
 msgctxt "#30031"
 msgid "Channels Logos Folder"
-msgstr "Cartella  loghi canali"
+msgstr "Cartella loghi canali"
```
Please remove this change. We only every update the en_gb strings.po file. All other translation files are done by transifix.

Be4stElectrjc replied:
Resolved.

phunkyfish reviewed (changes requested):
Ok, the changes look ok. You could move the header modification code to it's own function if you like, just a suggestion.

One thing though. The feature will override or add a user-agent protocol option to all channels. Is that really what most users will need? Really there should always be a way to prefer the user-agent in the m3u if there is one. What do you think? That might mean you need more than one setting.

Be4stElectrjc committed:
Added User-Agent strings

Be4stElectrjc commented:
Well, added the option to override the User-Agent in the playlist assuming the user wants total control on the addon. But i can remove it, less code. In this case, the feature will be used if no User-Agent is provided as header in the stream. Implemented the change pretty hardcoded, will move the code in a function to be used for headers handling, like in Matrix.

phunkyfish commented:
So what’s important is that the behaviour is the same when a user migrates from Leia to Matrix.

As there are already ways to override user-agent in Matrix you should figure out how you want the feature to work there.

I’m only giving you guidance, but always think of the user experience when adding features.

phunkyfish commented:
Ok, looks good. Let me know once you have the PR ready for Matrix. Once that's in we can look to merge this one.

This will also need a minor version increment in addon.xml and a changelog entry. No rush on this, just needs to happen before merge.

Be4stElectrjc committed:
Added header handling

Be4stElectrjc commented:
Have PR ready for Matrix.

Made modifications to Leia, no hardcoding and letting priority to the playlist like happens in Matrix.

phunkyfish commented:
Please open a new one. Just put matrix in The title ;)

phunkyfish commented on src/PVRIptvData.cpp:
```diff
@@ -1482,3 +1487,26 @@ int PVRIptvData::GetChannelId(const char * strChannelName, const char * strStrea
+  {
+    size_t present = strLine.find(header + '=', found + 1);
+
+    if (present == std::string::npos)
```
Did you forget to add the header here?

Be4stElectrjc replied:
Changed the variable to present, considering have header as argument of the function. The header it's added at the end.

phunkyfish commented on src/PVRIptvData.cpp:
```diff
@@ -1482,3 +1487,26 @@ int PVRIptvData::GetChannelId(const char * strChannelName, const char * strStrea
+      strLine += '&';
+    }
+
+    return strLine;
```
Why directly modify the reference and then return it?

Be4stElectrjc replied:
Because forgot putting the else.

phunkyfish commented on src/PVRIptvData.cpp:
```diff
@@ -1482,3 +1487,26 @@ int PVRIptvData::GetChannelId(const char * strChannelName, const char * strStrea
+    strLine += '|';
+  }
+
+  return strLine += header + '=' + g_userAgent;
```
This statement could just be in the else above as it will only be executed if the else is.

Be4stElectrjc replied:
This statement must be executed in two cases, so instead of writing it twice, i decided to write it once at the end.

phunkyfish replied:
Makes sense, although modifying the strLine variable and then returning the result doesn't make sense. I think it will confuse other devs. Maybe just return the value.

Be4stElectrjc replied:
That flow never happens. The return it's inside an else indicating the header was already present, so the playlist has priority over user setting. Before in the else block there was the code to erase the header from the current stream line, so it was more clear. Can add comments to the code if needed, not a prob.

phunkyfish replied:
But why does this make sense:
`return strLine += header + '=' + g_userAgent;`

Why not just just return `strLine + header + '=' + g_userAgent;`

Be4stElectrjc replied:
It's not a prob, used this one because it was committed by you on the Matrix branch.

phunkyfish commented on src/PVRIptvData.cpp:
```diff
@@ -802,7 +802,12 @@ bool PVRIptvData::LoadPlayList(void)
       channel.properties        = tmpChannel.properties;
-      channel.strStreamURL      = strLine;
+      
+      if (!g_userAgent.empty() && channel.properties.find("http-user-agent") == channel.properties.end())
```
What about also populating user-agent from the property if it's present?

Be4stElectrjc replied:
It's not needed because all properties from the stream are already taken here.

phunkyfish replied:
Yes, but nothing is done with the user-agent property. If just stays in the channel properties and is never used as far as I can see.

Be4stElectrjc replied:
It must be used, if not the properties mechanism wouldn't work at all. A thing that can be tested quickly with a playlist containing properties (didn't try).

phunkyfish replied:
All the properties mechanism does is make a key value pair available to the backend. It does not necessarily do anything with it. Certainly not a vlc only property.

Be4stElectrjc replied:
It's ok for the purpose. Since it's a key-value pair already available from the stream i can't delete or edit a property from the playlist, since has higher priority. The only thing i can do it's to use the setting when it's not available.

phunkyfish replied:
But then why not this:

```
auto propPair = channel.properties.find("http-user-agent");
if (propPair != channel.properties.end())
  channel.strStreamURL = AddHeaderToStreamUrl(strLine, "user-agent", propPair->second);
else (!g_userAgent.empty())
  channel.strStreamURL = AddHeaderToStreamUrl(strLine, "user-agent", g_userAgent);
```

Then you replicate the same functionality there is in Matrix.

Be4stElectrjc replied:
Yeah, meanwhile i could add that unused property of the playlist.

phunkyfish replied:
I think that is the unused property of the playlist 😉

Be4stElectrjc committed:
Updated readme

Be4stElectrjc commented:
> This will also need a minor version increment in addon.xml and a changelog entry. No rush on this, just needs to happen before merge.

All links in the addon.xml are dead btw, both here and in Matrix...

They should be correct for en_gb which are the only ones that matter.

phunkyfish commented on src/PVRIptvData.cpp:
```diff
@@ -1482,3 +1487,28 @@ int PVRIptvData::GetChannelId(const char * strChannelName, const char * strStrea
   return abs(iId);
 }
+
+std::string& PVRIptvData::AddHeaderToStreamUrl(std::string& strLine, const std::string& header)
```
Why not pass the strLine as a const url and just capture the separator char before the return?

So with a signature of:
`std::string PVRIptvData::AddHeaderToStreamUrl(const std::string& url, const std::string& header)`

phunkyfish replied:
And then a `char separator = '|';` which get's used in the return.

Be4stElectrjc replied:
Keep in mind the char separator depends about there are or not other headers in the stream line.
For the signature i would keep strLine to be more clear, even if not strictly necessary (opinion). Used return by reference to optimize speed, but for a built-in type it's negligible.

phunkyfish replied:
Ok, but then why return anything at all? Just modify strLine OR return a value. Don’t do both.

Also your function is called add header to stream url but it’s argument is called strLine. Variable names should be relevant to the local scope, not the parent scope.

Be4stElectrjc replied:
Yeah, just saw it. 
You're right, it's better with that function name.

phunkyfish commented on src/PVRIptvData.cpp:
```diff
@@ -1482,3 +1487,28 @@ int PVRIptvData::GetChannelId(const char * strChannelName, const char * strStrea
+
+  if (found != std::string::npos)
+  {
+    size_t present = strLine.find(header + '=', found + 1);
```
Why not put the find call in the if statement as you don't use the `present` variable other than in the if.

Be4stElectrjc replied:
I can save one variable and a line using it directly inside the if statement, for sure. The purpose it's to make the lines shorter.

phunkyfish commented on src/PVRIptvData.cpp:
```diff
@@ -1482,3 +1487,28 @@ int PVRIptvData::GetChannelId(const char * strChannelName, const char * strStrea
+    size_t present = strLine.find(header + '=', found + 1);
+
+    if (present == std::string::npos)
+    {
```
You don't need either set of brackets in this if/else.

Be4stElectrjc replied:
I need the return only to be executed when the if condition it's not verified. Not in both cases. However, i can spare some brackets for the inner "single line" if/else statements.

phunkyfish commented:
Here is what I mean about the signature, return type and use a separator char instead. Now the function could be used for adding any header.

```
std::string PVRIptvData::AddHeaderToStreamUrl(const std::string& url, const std::string& header, const std::string& value)
{
  size_t found = strLine.find('|');

  char separator = '|';
  if (found != std::string::npos)
  {
    if (strLine.find(header + '=', found + 1) != std::string::npos)
      return url;
    else
      separator = '&';
  }

  return url + separator + header + '=' + value;
}
```

Be4stElectrjc commented:
I tried to remain consistent with your function in Matrix and be explicit, but you saved my time optimizing it. Nice work.

`const std::string& PVRIptvData::AddHeaderToStreamUrl(const std::string& url, const std::string& header, const std::string& value) const`

The const in the return because we return a const, the const for the function as caution.
Changes on the go.

phunkyfish commented:
Ok, and then you just need addon.xml and changelog similar to Matrix. I think on Leia the changelog entry is only in the changelog.

Be4stElectrjc committed:
Updated version

phunkyfish commented on src/PVRIptvData.cpp:
```diff
@@ -1482,3 +1492,23 @@ int PVRIptvData::GetChannelId(const char * strChannelName, const char * strStrea
   return abs(iId);
 }
+
+const std::string& PVRIptvData::AddHeaderToStreamUrl (const std::string& url, const std::string& header, const std::string& value) const
```
The return type should be `std:string` and not `std::string&` as the returned variable will be on the stack. 

The compiler will optimise so the object is only created once anyway and it will act like a returned reference.

phunkyfish commented on src/PVRIptvData.cpp:
```diff
@@ -802,7 +802,17 @@ bool PVRIptvData::LoadPlayList(void)
       channel.iTvgShift         = tmpChannel.iTvgShift;
       channel.bRadio            = tmpChannel.bRadio;
       channel.properties        = tmpChannel.properties;
-      channel.strStreamURL      = strLine;
```
You need to keep this line. Otherwise you will never set the stream url if you are not setting user agent.

Be4stElectrjc replied:
`else {channel.strStreamURL = strLine;}`

phunkyfish replied:
That will work too.

phunkyfish reviewed (changes requested):
Just the bug above to fix and then I think this is good.

phunkyfish commented on src/PVRIptvData.cpp:
```diff
@@ -802,7 +802,21 @@ bool PVRIptvData::LoadPlayList(void)
+      {
+        channel.strStreamURL = AddHeaderToStreamUrl(strLine, "user-agent", g_userAgent);
+      }
+	  else
```
Indentation is wrong. Should be inline with rest of statement.

Be4stElectrjc replied:
Now it's correct.

phunkyfish reviewed (approved):
Nice. Well done for putting up with my reviews 😉

Thanks again for contributing!

phunkyfish commented:
Oops, we actually got the version number wrong. It should have been 3.9.0!

No harm done though 😉

Be4stElectrjc commented:
You can always correct it, ty for the help.

phunkyfish commented:
I think it’s ok as it is. I don’t think there will be another Leia release before Matrix.

Thanks again!
